presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java

```diff
@@ -378,18 +420,18 @@ public void setMemoryReservation(long newMemoryReservation)
         }
     }
 
-    public boolean trySetMemoryReservation(long newMemoryReservation)
+    public boolean tryReserveMemory(long newMemoryReservation)
     {
         checkArgument(newMemoryReservation >= 0, "newMemoryReservation is negative");
 
-        long delta = newMemoryReservation - memoryReservation.get();
+        long delta = newMemoryReservation - operatorMemoryContext.reservedUserMemory();
 
         if (delta > 0) {
             if (!driverContext.tryReserveMemory(delta)) {
                 return false;
             }
 
-            memoryReservation.addAndGet(delta);
+            operatorMemoryContext.reserveUserMemory(delta);
```

Comment:
This is quite different from the old code although they look somewhat similar. Originally, we traverse to the top once. We are now traversing to the top twice. I would make it possible to do the same with the new MemoryContext structure. However, I understand that doing this alone doesn't help. The top memory context must be able to push back. This is possible with the original inheritance mechanism. But I'm not sure whether this is possible with the listener mechanism.